data/calculated_parser/functions.py

```diff
@@ -69,11 +70,293 @@ def sspeed(depth, latitude, temperature, salinity):
         press = [seawater.pres(d, latitude) for d in depth]
     except TypeError:
         press = seawater.pres(depth, latitude)
-
+    
     speed = seawater.svel(salinity, temperature, press)
     return np.array(speed)
 
 
+def sspeedmin(depth, lat, lon, temperature, salinity):
+    """
+    Finds the global minimum of the speed of sound
+
+    Parameters:
+    sspeed: Speed of Sound
+    latitude: The latitude(s) in degrees North
+    """
+    
+    speed = sspeed(depth, lat, temperature, salinity)
+    speed = speed.transpose()
+    
+    for x in range(speed.shape[0]):
+        for y in range(speed.shape[1]):
+            min_val = np.nanmin(speed[x,y])
+            idx = np.where(speed[x,y] == min_val)
+            if (np.isnan(min_val)):
+                speed[x,y] = np.nan
+            elif idx[0].shape[0] > 1:
+                idx = idx[0][0]
+
+                speed[x,y] = depth.values[idx]
+            else:
+                speed[x,y] = depth.values[idx]
+            #speed[x][y] = depth[np.where(speed[x][y] == np.nanmin(speed[x][y]))]  #np.nanmin(speed[x][y])
+    
+    speed = speed.transpose()
+
+    speed = speed[0]
+
+    return np.array(speed)
+
+def soniclayerdepth(depth, lat, lon, temperature, salinity):
+    """
+    Finds the local maxima of the speed of sound
+
+    Parameters:
+    sspeed: Speed of Sound
+    latitude: The latitude(s) in degrees North
+    """
+    
+    speed = sspeed(depth, lat, temperature, salinity)
+    speed = speed.transpose()
+    sld = np.nan
+    for x in range(speed.shape[0]):
+        for y in range(speed.shape[1]):
+            sca_value = np.nanmin(speed[x,y])
+            sca_idx = np.where(speed[x,y] == sca_value)
+            
+            if (np.isnan(sca_value)):
+                pass
+            else:
+                sca_idx = sca_idx[0][0]
+
+                subset = speed[x,y][0:int(sca_idx) + 1]
+                sld_value = subset.max()
+                
+                if (np.isnan(sld_value)):
+                    pass
+                else:
+                    sld_idx = np.where(subset == sld_value)[0][0]
+                    sld = depth.values[sld_idx]
+                    speed[x,y] = sld
+
+    speed = speed.transpose()
```

Comment:
Same comment as above for double transpose.

Reply:
This was part of my initial attempts at making it work. I'm not sure what the transpose does at all... I'll investigate on monday to see if it's needed

Reply:
If we know the speed is shape [z,y,x] then we can avoid the transpose by iterating like this:
```python
sld = np.nan((speed.shape[-2], speed.shape[-1]))
for x in range(speed.shape[-1]):
   for y in range(speed.shape[-2]):
      sca_value = np.nanmin(speed[:, y, x])
      ...
      sld[y, x] = ...
return sld
```

But, it would take more thought to make sure it works for speed arrays that are shape [t,z,y,x]  if this scenario is possible. ( I'm not sure how the time series functionality works in the calculation layer). In fact, I'm not sure the transpose works for arrays shape [t,z,...] either. Has that been tested?

From what I see the transpose moves the depth dimension from the beginning to the end, for example [z,y,x] to [y,x,z].

Reply:
So I do remember the transpose, and it was important for the way I was doing it. However, Nancy's solution also works. I will convert everything over to using this method.

Reply:
removed double transpose using nancy's solution